build: Avoid `BOOST_NO_CXX98_FUNCTION_BASE` macro redefinition

hebasto opened the pull request:
With GCC 12 and Boost 1.81 (from depends) having multiple warnings:
```
In file included from /home/hebasto/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/config.hpp:48:
/home/hebasto/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/config/stdlib/libstdcpp3.hpp:397:9: warning: 'BOOST_NO_CXX98_FUNCTION_BASE' macro redefined [-Wmacro-redefined]
#define BOOST_NO_CXX98_FUNCTION_BASE
        ^
<command line>:8:9: note: previous definition is here
#define BOOST_NO_CXX98_FUNCTION_BASE 1
        ^
1 warning generated.
```

This PR fixes those warnings.

Defining of the `BOOST_NO_CXX98_FUNCTION_BASE` macro was introduced in https://github.com/bitcoin/bitcoin/pull/25436, but since https://github.com/boostorg/config/pull/430, it is required to check it before adding.

fanquake commented on configure.ac:
```diff
@@ -1479,9 +1479,23 @@ if test "$use_boost" = "yes"; then
+  dnl and will generate warnings with newer compilers for Boost
+  dnl older then 1.80.
+  dnl See:
+  dnl   - https://github.com/boostorg/container_hash/issues/22
```
We only need to link to a single upstream issue, probably https://github.com/boostorg/config/pull/430.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26952#issuecomment-1401808126).

hebasto commented:
Updated 387a99ddb2fd2da4e37a99b370eea5ae6df64c68 -> bb76ac411c45991ace51e14f2c685d138ca146bf ([pr26952.01](https://github.com/hebasto/bitcoin/commits/pr26952.01) -> [pr26952.02](https://github.com/hebasto/bitcoin/commits/pr26952.02), [diff](https://github.com/hebasto/bitcoin/compare/pr26952.01..pr26952.02)):

- addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/26952#discussion_r1085145547)

fanquake commented:
I don't think this works? `BOOST_NO_CXX98_FUNCTION_BASE` is still set when using Boost 1.81.0 and GCC 12. So you still get the warnings. i.e with this branch (rebased on master):
```bash
# gcc-12 (Ubuntu 12.1.0-2ubuntu1~22.04) 12.1.0
make -C depends/ CC=gcc-12 CXX=g++-12
./autogen.sh
CONFIG_SITE=/home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/share/config.site ./configure
checking whether BOOST_NO_CXX98_FUNCTION_BASE macro definition is needed... yes
make -j9
...
In file included from /home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/config.hpp:48,
                 from /home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/multi_index/hashed_index.hpp:16,
                 from ./txmempool.h:35,
                 from blockencodings.cpp:13:
/home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/config/stdlib/libstdcpp3.hpp:397: warning: "BOOST_NO_CXX98_FUNCTION_BASE" redefined
  397 | #define BOOST_NO_CXX98_FUNCTION_BASE
      | 
```

hebasto committed:
build: Avoid `BOOST_NO_CXX98_FUNCTION_BASE` macro redefinition

hebasto commented:
> I don't think this works?

Thanks! Should be fixed now.

fanquake reviewed (approved):
ACK d4c59da8d6e3aac14306249aa12332fed55efebd - it works now.
